cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -122,15 +162,23 @@ public void setAuditPublisher(AuditPublisher auditPublisher) {
    * @throws NotFoundException if either the namespace or dataset instance is not found,
    * @throws IOException if there is a problem in making an HTTP request to check if the namespace exists.
    */
-  public DatasetMeta get(Id.DatasetInstance instance, List<? extends Id> owners) throws Exception {
+  DatasetMeta get(final Id.DatasetInstance instance, List<? extends Id> owners) throws Exception {
+    // Application Deployment first makes a call to the dataset service to check if the instance already exists with
+    // a different type. To make sure that that call responds with the right exceptions if necessary, first fetch the
+    // meta from the cache and throw appropriate exceptions if necessary.
```

Comment:
If someone doesn't have access, it shouldn't be able to tell if a dataset exists or not.